direct.go

```diff
@@ -311,78 +316,88 @@ func cloneRequestWith(req *http.Request, frontedHost string, body io.ReadCloser)
 }
 
-func (d *direct) dialWith(in chan masquerade) (net.Conn, *masquerade, func(bool) bool, bool, error) {
+func (d *direct) dialWith(vetted chan masquerade, cached chan masquerade) (net.Conn, *masquerade, func(bool) bool, bool, error) {
 	retryLater := make([]masquerade, 0)
 	defer func() {
+		log.Debugf("%d masquerades to retry", len(retryLater))
+		log.Debugf("capacity %d", len(vetted))
 		for _, m := range retryLater {
 			// when network just recovered from offline, retryLater has more
 			// elements than the capacity of the channel.
 			select {
-			case in <- m:
+			case vetted <- m:
 			default:
 				log.Debug("Dropping masquerade: retry channel full")
 			}
 		}
 	}()
 
 	for {
+		// order of preference vetted -> cached -> unvetted
 		var m masquerade
 		select {
-		case m = <-in:
+		case m = <-vetted:
 			log.Trace("Got vetted masquerade")
 		default:
-			log.Trace("No vetted masquerade found, falling back to unvetted candidate")
 			select {
-			case m = <-d.candidates:
-				log.Trace("Got unvetted masquerade")
+			case m = <-cached:
+				log.Trace("Got cached masquerade")
 			default:
-				return nil, nil, nil, false, errors.New("Could not dial any masquerade?")
+				log.Trace("No vetted or cached masquerade found, falling back to unvetted candidate")
+				select {
+				case m = <-d.candidates:
+					log.Trace("Got unvetted masquerade")
+				default:
+					return nil, nil, nil, false, errors.New("Could not dial any masquerade?")
+				}
 			}
 		}
 
-		log.Tracef("Dialing to %v", m)
+		log.Debugf("Dialing to %v", m)
 
 		// We do the full TLS connection here because in practice the domains at a given IP
 		// address can change frequently on CDNs, so the certificate may not match what
 		// we expect.
 		conn, retriable, err := d.doDial(&m.Masquerade)
+		masqueradeGood := func(good bool) bool {
+			if good {
+				m.LastVetted = time.Now()
+				// Requeue the working connection to masquerades
+				d.masquerades <- m
+				select {
+				case d.toCache <- &cacheOp{m: m}:
+					// ok
+				default:
+					// cache writing has fallen behind, drop masquerade
+					log.Debug("Dropping masquerade: cache writing is behind")
+				}
+			} else {
+				d.toCache <- &cacheOp{m: m, remove: true}
+			}
+
+			return good
+		}
 		if err == nil {
 			log.Trace("Returning connection")
-			masqueradeGood := func(good bool) bool {
-				if good {
-					m.LastVetted = time.Now()
-					// Requeue the working connection to masquerades
-					d.masquerades <- m
-					select {
-					case d.toCache <- m:
-						// ok
-					default:
-						// cache writing has fallen behind, drop masquerade
-						log.Debug("Dropping masquerade: cache writing is behind")
-					}
-				} else {
-					go d.vetOneUntilGood()
-				}
 
-				return good
-			}
 			return conn, &m, masqueradeGood, true, err
 		} else if retriable {
 			retryLater = append(retryLater, m)
 		} else {
 			log.Debugf("Dropping masquerade: non retryable error: %v", err)
+			masqueradeGood(false)
```

Comment:
On non-retriable error, mark masquerade as bad so that we vet another.

Reply:
I notice we used to proactively kick off a vetOneUntilGood() in a goroutine from masqueradeGood(false). I'm not sure where this new vetting is triggered now.

Are we relying on dialWith eventually/lazily selecting unvetted items to do the vetting here?

Reply:
Oh good catch, that needs to still happen. Fixed.

Reply:
might be a good candidate for a new test of some sort, though I think it would have still muddled through heh.